distributed/shuffle/tests/test_shuffle.py

```diff
@@ -1481,8 +1354,274 @@ async def receive(self, data: list[bytes]) -> None:
         loop=loop_in_thread,
     )
     try:
-        await sB.add_partition(dfs[0])
+        await sB.add_partition(dfs[0], 0)
         with pytest.raises(RuntimeError, match="Error during receive"):
             await sB.barrier()
     finally:
         await asyncio.gather(*[s.close() for s in [sA, sB]])
+
+
+from distributed.worker import DEFAULT_EXTENSIONS
+
+
+class BlockedShuffleReceiveShuffleWorkerExtension(ShuffleWorkerExtension):
+    def __init__(self, worker: Worker) -> None:
+        super().__init__(worker)
+        self.in_shuffle_receive = asyncio.Event()
+        self.block_shuffle_receive = asyncio.Event()
+
+    async def shuffle_receive(self, *args: Any, **kwargs: Any) -> None:
+        self.in_shuffle_receive.set()
+        await self.block_shuffle_receive.wait()
+        return await super().shuffle_receive(*args, **kwargs)
```

Comment:
I don't love the mocking we're doing all the time for these rather primitive hooks. I wonder if we should add a couple of common hooks we can set in tests, e.g. `_prereceive_hook`, etc.

not that important right now, just a thought